docs: add navigation failures (#3220)
vuejs/vue-router: docs/guide/advanced/navigation-failures.md

# Navigation Failures

When using `router-link`, Vue Router internally calls `router.push` to trigger a navigation. Depending on the current location and existing [Navigation Guards](./navigation-guards.md), this navigation might end up in a new page being shown, but there are a couple of situations where we will stay on the same page:

- We are already on the page we are trying to go to
- A navigation guard aborts the navigation by calling `next(false)`

> - An error is thrown in a navigation by calling `next(new Error())`

Author: ```suggestion
- A navigation guard throws an error or calls `next(new Error())`
```

> When using a regular `router-link`, **none of these failures will log an error**. However, if you are using `router.push` or `router.replace`, you might come across an _"Uncaught (in promise) Error"_ message followed by a more specific message in your console. This is part of the _Navigation Failures_ system in Vue Router and it is only available from version 3.2.0 onwards but existed for a long time before: through the two optional callbacks, `onComplete` and `onAbort` that can be passed to `router.push`. Since version 3.1.0, `router.push` and `router.replace` return a _Promise_ if no `onComplete`/`onAbort` callback is provided. This _Promise_ resolves instead of invoking `osComplete` and rejects instead of invoking `onAbort`.

Reviewer: You mention a "regular" `router-link`, but is there a special `router-link` that I should be aware of?

Reviewer: > **none of these failures will log an error**

Will these failures still cause an error even if they're not logged?

Author: You are right, it makes no sense to talk about regular. I was thinking about using a custom router-link

Navigation failures in router-link will not cause an error either, we pass an empty `onAbort` handler to silence them

> When using a regular `router-link`, **none of these failures will log an error**. However, if you are using `router.push` or `router.replace`, you might come across an _"Uncaught (in promise) Error"_ message followed by a more specific message in your console. This is part of the _Navigation Failures_ system in Vue Router and it is only available from version 3.2.0 onwards but existed for a long time before: through the two optional callbacks, `onComplete` and `onAbort` that can be passed to `router.push`. Since version 3.1.0, `router.push` and `router.replace` return a _Promise_ if no `onComplete`/`onAbort` callback is provided. This _Promise_ resolves instead of invoking `osComplete` and rejects instead of invoking `onAbort`.

Reviewer 2: Typo 😉 

```suggestion
When using a regular `router-link`, **none of these failures will log an error**. However, if you are using `router.push` or `router.replace`, you might come across an _"Uncaught (in promise) Error"_ message followed by a more specific message in your console. This is part of the _Navigation Failures_ system in Vue Router and it is only available from version 3.2.0 onwards but existed for a long time before: through the two optional callbacks, `onComplete` and `onAbort` that can be passed to `router.push`. Since version 3.1.0, `router.push` and `router.replace` return a _Promise_ if no `onComplete`/`onAbort` callback is provided. This _Promise_ resolves instead of invoking `onComplete` and rejects instead of invoking `onAbort`.
```

## Detecting Navigation Failures

_Navigation Failures_ are `Error` instances with a few extra properties. Among them, you can find a `type` property. This will allow you to check the type of the navigation failure:

```js

> import { NavigationFailureTypes } from 'vue-router'

Author: I was thinking we could also expose an `isRouterError` function:

```js
router.push('/admin').catch(failure => {
   if (isRouterError(failure)) // any router error
   if (isRouterError(failure, NavigationFailureTypes.abort)) // only aborted navigations
})
```

> import { NavigationFailureTypes } from 'vue-router'

Reviewer 3: As far as I can see, you're not actually exporting this? Importing it just gives me `undefined` with v3.3.4, and it's not present in the typescript types either.

Author: This is not yet implemented in vue router 3

Reviewer 3: Okay, I think removing the name before implementing this was a bad idea, but I guess you've gathered as much from the comments in #3047.

I think checking the error type as documented in these docs is fine when `NavigationFailureTypes` is exported. However, as far as I can see, you silently ignore the types you don't check. You should rethrow the error if it's any of the other types.

Author: Yeah, I think will add back the `name` for the only error that exposed a `name` before but it won't be documented anywhere.
It's true the examples could show how a non caught error is thrown

# Navigation Failures

When using `router-link`, Vue Router internally calls `router.push` to trigger a navigation. Depending on the current location and existing [Navigation Guards](./navigation-guards.md), this navigation might end up in a new page being shown, but there are a couple of situations where we will stay on the same page:

- We are already on the page we are trying to go to
- A navigation guard aborts the navigation by calling `next(false)`
- A navigation guard throws an error or calls `next(new Error())`

> When using a regular `router-link`, **none of these failures will log an error**. However, if you are using `router.push` or `router.replace`, you might come across an _"Uncaught (in promise) Error"_ message followed by a more specific message in your console. This is part of the _Navigation Failures_ system in Vue Router and it is only available from version 3.2.0 onwards but existed for a long time before: through the two optional callbacks, `onComplete` and `onAbort` that can be passed to `router.push`. Since version 3.1.0, `router.push` and `router.replace` return a _Promise_ if no `onComplete`/`onAbort` callback is provided. This _Promise_ resolves instead of invoking `onComplete` and rejects instead of invoking `onAbort`.

Reviewer 4: Would it make sense to split this into 2-3 paragraphs? 
1) About the error handling when using `router-link`.
2) Using the `onComplete` and `onAbort` callbacks.
"New in 3.1.0" tag
3) `router.push` and `router.replace` return a _Promise_.

Author: Since the Navigation failure system is only available from 3.4.0 (the next release), I don't think it makes much sense to add a new in 3.1.0 but instead we could put a New in 3.4.0 at the very top

How would you split into multiple paragraphs the explanation? They are all related sentences

Author: I think I should remove the whole reference to versions if I add something at the very top

Author: The thing is people often want to know why things are like that, that's why I added the version things were added at

Reviewer 4: That makes sense. I wonder though if it is a good practice to mix historical context with informative docs. As in, the historical context might be very interesting to read but likely shouldn’t be required to understand what is happening. Maybe a ::tip:: with informations of how it changed across different versions?

Author: That's true, I updated that part, let me know what you think

Reviewer 4: Looks great! Thank you!

> ## Detecting Navigation Failures

Reviewer 4: I assume this is introduced with v3.2.0 acccording to the previous paragraph. What about adding a tag `New in 3.2.0`?

_Navigation Failures_ are `Error` instances with a few extra properties. To check if an error comes from the Router, use the `isRouterError` function:

```js

> import { NavigationFailureType, isRouterError } from 'vue-router'

Author: Maybe something like `isNavigationFailure` is better naming than `isRouterError` as there aren't any other errors and it's also confusing for it to be named error and failure

Reviewer 4: Agree. I would also lean towards `isRoutingError` given that this is an `Error` instance. So either `isNavigationError` or `isRoutingError`.

Author: Oh I was going the other way around, it's an Error instance to have a stack trace but it's technically not an error as it's expected behavior. Maybe Failure isn't a good name either

Reviewer 4: What about `Rejection`?

Reviewer 4: It’s much softer than both `Error` and `Failure` and might be quite good at communicating what is actually happening, as in this navigation request has been rejected. Not an error (this is the expected behaviour/feature). This would also allow us to actually have Errors thrown that are actual errors (as in unexpected).

Author: Yeah, it's softer but it's also related to Promise rejection and in v4, they appear when the promise resolve. It would also require changing the `NavigationRejectionType` name, but that could only be done in v4 and add an amendment to the RFC that was merged

Reviewer 4: That makes sense! In that case maybe Exception? Running out of ideas here :(

Author: There is also NavigationDropped, Canceled (one of the types), Stopped

Reviewer 4: In that case, I think Failure is a good word.

// trying to access the admin page
router.push('/admin').catch(failure => {

>   if (failure) {

Reviewer 4: Is there a case where this would be _falsy_ even though the `catch` was triggered?

Author: No, that shouldn't happen, so I think we can remove this!

Reviewer 4: Great!

# Navigation Failures

> New in 3.4.0

> When using `router-link`, Vue Router internally calls `router.push` to trigger a navigation. Depending on the current location and existing [Navigation Guards](./navigation-guards.md), this navigation might end up in a new page being shown, but there are a couple of situations where we will stay on the same page:

Reviewer: ```suggestion
When using `router-link`, Vue Router calls `router.push` to trigger a navigation. While the expected behavior for most links is to navigate a user to a new page, there are a few situations where users will remain on the same page:
```

> - We are already on the page we are trying to go to

Reviewer: ```suggestion
- Users are already on the page that they are trying to navigate to
```

Author: Should we say _that is being navigated to_? as there isn't any request going on or is it still clearer that way?

Reviewer: Updated!

> - A navigation guard aborts the navigation by calling `next(false)`
> - A navigation guard throws an error or calls `next(new Error())`

Reviewer: ```suggestion
- A [navigation guard](./navigation-guards.md) aborts the navigation by calling `next(false)`
- A [navigation guard](./navigation-guards.md) throws an error or calls `next(new Error())`
```

> When using a regular `router-link`, **none of these failures will log an error**. However, if you are using `router.push` or `router.replace`, you might come across an _"Uncaught (in promise) Error"_ message followed by a more specific message in your console. Let's understand how to differentiate _Navigation Failures_.

Reviewer: ```suggestion
When using a `router-link` component, **none of these failures will log an error**. However, if you are using `router.push` or `router.replace`, you might come across an _"Uncaught (in promise) Error"_ message followed by a more specific message in your console. Let's understand how to differentiate _Navigation Failures_.
```

Author: should just omit it? or maybe say _component_?

Reviewer: Works for me!

::: tip Background story

>  _Navigation Failures_ were exposed on version 3.2.0 but existed for a long time before: through the two optional callbacks of `router.push`, `onComplete` and `onAbort` that can be passed to `router.push`. Since version 3.1.0, `router.push` and `router.replace` return a _Promise_ if no `onComplete`/`onAbort` callback is provided. This _Promise_ resolves instead of invoking `onComplete` and rejects instead of invoking `onAbort`.

Reviewer: ```suggestion
 In v3.2.0, _Navigation Failures_ were exposed through the two optional callbacks of `router.push`: `onComplete` and `onAbort`. Since version 3.1.0, `router.push` and `router.replace` return a _Promise_ if no `onComplete`/`onAbort` callback is provided. This _Promise_ resolves instead of invoking `onComplete` and rejects instead of invoking `onAbort`.
```

 :::

## Detecting Navigation Failures

_Navigation Failures_ are `Error` instances with a few extra properties. To check if an error comes from the Router, use the `isNavigationFailure` function:

```js
import { NavigationFailureType, isNavigationFailure } from 'vue-router'

// trying to access the admin page
router.push('/admin').catch(failure => {
  if (isNavigationFailure(failure, NavigationFailureType.redirected)) {
    // show a small notification to the user
    showToast('Login in order to access the admin panel')
  }
})
```

::: tip
If you omit the second parameter: `isNavigationFailure(failure)`, it will only check if the error is a _Navigation Failure_.
:::

## `NavigationFailureType`

> `NavigationFailureType` exposes the following properties to differentiate _Navigation Failures_:

Reviewer: ```suggestion
`NavigationFailureType` help developers to differentiate between the various types of _Navigation Failures_. There are four different types:
```

- `redirected`: `next(newLocation)` was called inside of a navigation guard to redirect somewhere else.
- `aborted`: `next(false)` was called inside of a navigation guard to the navigation.
- `cancelled`: A new navigation completely took place before the current navigation could finish. e.g. `router.push` was called while waiting inside of a navigation guard.
- `duplicated`: The navigation was prevented because we are already at the target location.

## _Navigation Failures_'s properties

All navigation failures expose `to` and `from` properties to reflect the current location as well as the target location for the navigation that failed:

```js
// trying to access the admin page
router.push('/admin').catch(failure => {
  if (isNavigationFailure(failure, NavigationFailureType.redirected)) {
    failure.to.path // '/admin'
    failure.from.path // '/'
  }
})
```

> In all cases, `from` and `to` are normalized route locations.

Reviewer: ```suggestion
In all cases, `to` and `from` are normalized route locations.
```
